Jetpack Cloud: Use separate environments

tyxla opened the pull request:
This PR is an experiment to make the Jetpack Cloud work as a separate development and production instance by taking advantage of new Calypso environments. Here we're introducing 3 new environments:

* `jetpack-cloud-development` - Local, development environment. This is the Jetpack Cloud version of the regular Calypso development environment.
* `jetpack-cloud-stage` - Staging environment. This is the Jetpack Cloud version of the regular Calypso staging environment, which is close to the production one but used for testing (when proxied) before deploying to production.
* `jetpack-cloud-production` - Production environment. This is the Jetpack Cloud version of the regular Calypso production environment. 

Those will be accessible in their corresponding domains, as follows:

* `jetpack-cloud-development` accessible locally on http://jetpack.cloud.localhost:3000/
* `jetpack-cloud-stage` accessible on https://cloud.jetpack.com/ (when proxied)
* `jetpack-cloud-production` accessible on https://cloud.jetpack.com/ (when not proxied)

By using Calypso environments we can natively take advantage of all the Calypso features, use the existing development features, staging and production, and deployment infrastructure.

Note that the proxied/non-proxied logic will have to be set up additionally, likely with the help of systems.

#### Changes proposed in this Pull Request

* Introduce 3 new environments for Jetpack Cloud: `development`, `stage` and `production`.
* Remove base route prefix for Jetpack Cloud to make the routes accessible from the root.
* Update server pages to load the Jetpack Cloud entry point for those environments

#### Testing instructions

* Checkout this branch.
* Point `jetpack.cloud.localhost:3000` to `127.0.0.1` in your hosts file.
* Point `cloud.jetpack.com` to `127.0.0.1` in your hosts file.
* Run `CALYPSO_ENV=jetpack-cloud-development npm start`
* Go to http://jetpack.cloud.localhost:3000/
* Verify you can access the page and you can navigate to the other pages.
* Go to http://jetpack.cloud.localhost:3000/scan directly and verify you can access it.
* Stop the process.
* Make sure you have Docker installed and it's currently running.
* Make sure you're setup Docker to use 8G of memory and add some CPUs if possible (as documented in FG here: PCYsg-5XR-p2)
* Run `npm run build-docker` - note that it takes time (15 to 30 min) and takes over your machine 😉 
* When it finishes, run: `docker run -it --env CALYPSO_ENV=jetpack-cloud-production --name wp-calypso --rm -p 80:3000 wp-calypso`
* Go to http://cloud.jetpack.com/
* Verify you can access the page and you can navigate to the other pages.
* Note that because this is a subdomain of jetpack.com, testing it may not work for the first time. You may have to try the following:
	* Your browser might be forcing usage of https. Verify you're running http and not https.
	* Delete any browser cache (including the domain one on chrome://net-internals/#hsts).
	* Delete your system's DNS cache, and/or force reconnection if using autoproxxy.
	* Restart your browser.
	* Try with and without a proxy.
	* Try in incognito mode.
	* Try in another browser.
* Go to http://cloud.jetpack.com/scan directly and verify you can access it.
* Stop the process.
* Run `npm start`.
* Smoke test Calypso (http://calypso.localhost:3000/) and verify it still works well.
* Bonus: run production calypso in your local Docker instance and verify it works: 
  * Point `wpcalypso.wordpress.com` to `127.0.0.1` in your hosts file.
  * Delete the domain security policies on chrome://net-internals/#hsts if using Chrome.
  * Run `docker run -it --env CALYPSO_ENV=production --name wp-calypso --rm -p 80:3000 wp-calypso`.
  * Open http://wpcalypso.wordpress.com/ (note the HTTP) and verify it works.

#### Notes

Note that the commit for making staging/production http-only will be removed prior to merging. It's been added only to facilitate easier testing.

ChaosExAnima commented on client/server/pages/index.js:
```diff
@@ -645,6 +651,13 @@ module.exports = function() {
 	app.use( setupLoggedInContext );
 	app.use( handleLocaleSubdomains );
 
+	if ( jetpackCloudEnvs.includes( process.env.CALYPSO_ENV ) ) {
```
Is this always `CALYPSO_ENV`? I sometimes see `NODE_ENV`. Perhaps it would be more flexible to check the `config` env instead of directly reading the process?

tyxla replied:
Really good catch @ChaosExAnima! Fixed in https://github.com/Automattic/wp-calypso/pull/39058/commits/6a4d6750c185c560ae92f83fb3b941aed596d581

ChaosExAnima commented:
Hey @tyxla, dev works great but I'm having issues with testing stage:
<img width="1139" alt="Screen Shot 2020-01-24 at 11 31 51 AM" src="https://user-images.githubusercontent.com/1760168/73085759-38166200-3e9d-11ea-9a34-f8a92a8db928.png">

It loads, but the router seems to be having issues?

ChaosExAnima commented:
Okay, so did some tinkering. Unfortunately I couldn't get the Docker environment up and running, it just timed out with the build. And while the stage/prod environments don't work for me, I also note the normal Calypso ones fail with the same issues.

At this point, I'm going to defer to someone else to test environments outside of dev. But that works great for me!

enejb commented on client/landing/jetpack-cloud/section.js:
```diff
@@ -1,6 +1,6 @@
 export const JETPACK_CLOUD_SECTION_DEFINITION = {
 	name: 'jetpack-cloud',
-	paths: [ '/jetpack-cloud' ],
+	paths: [ '/', '/scan', '/backups' ],
```
Can we define this inside client/sections.js instead?

delawski replied:
Wouldn't `paths: [ '/' ]` alone suffice? Why do we have to explicitly define `/scan` and `/backups`?

tyxla replied:
> Can we define this inside client/sections.js instead?

Yes, but we're consciously not doing it right now because we're creating Jetpack Cloud only with this section to keep it clean. In the future, we could be combining those if we decide that we want to expose Calypso routes for Jetpack Cloud. So let's keep it as-is for this first version and iterate later.

tyxla replied:
> Wouldn't `paths: [ '/' ]` alone suffice? Why do we have to explicitly define `/scan` and `/backups`?

No, they won't. This would work if we had the entire Jetpack Cloud in a subroute, but because it is the landing route itself, then it needs all the first level routes to be defined so Express.js would know about them and would handle them properly.

tyxla commented:
Thanks for the feedback everyone! I'm iterating a bit before marking it as "ready to review" again.

tyxla commented:
> It loads, but the router seems to be having issues?

Please, give it a try with the latest testing instructions. The previous ones were very wrong, sorry for that 🤦‍♂

tyxla commented:
> One thing that I would suggest is adding a npm command that lets us start jetpack cloud easier.
Such as npm run start-jetpack for example

I like it! Let's build it in another PR once we land this one - it doesn't seem to be related with this PR.

tyxla commented:
Folks, I've addressed all the feedback and replied to all your questions.

Also, I've updated the testing instructions and verified that they work.

I'd really appreciate it if I can you to test and review the PR again. Thanks a bunch!

enejb commented:
I tested this as well and it work well. 

One thing that I noticed is that the we don't have a 404 page any more. For pages that don't exist yet.

tyxla commented:
> One thing that I noticed is that the we don't have a 404 page any more. For pages that don't exist yet.

Yeah, this is on purpose - I assumed we will want to build a custom one for the Jetpack Cloud. Can be built in another PR.

tyxla commented:
Thank you for the reviews and testing, folks!
